### DISCUSSION TITLE
Converting raw raster data to PNG using palette

### DISCUSSION THREAD

**Discussion by chickazee4** [ImageMagick/ImageMagick#5741] [Help] [answered]
Basically, there is this very old global digital elevation model from NOAA (https://www.ngdc.noaa.gov/mgg/topo/gltiles.html) that I would like to be able to programmatically convert (using a shell script, preferably) into a usable image without losing any fidelity. NOAA provides instructions for some specialized programs and what I can only presume to be an ancient version of Photoshop (https://www.ngdc.noaa.gov/mgg/topo/report/s11/s11G.html), but I do not have immediate access to any of the tools it references, so I'm trying to figure out a way to do this in ImageMagick. 

The model's data is provided as a set of 10800x4800px or 10800x6000px set of multiple "tiles," which are .RAW format raster images covering rectangular sections of a map of the Earth. A palette for all of the tiles is provided alongside them in both .ACT and .PAL format. I'd like to convert each "tile" into a colored .PNG image (or something equivalent) so that I can combine the resulting files into a single image resembling [the one on the model's home page](https://www.ngdc.noaa.gov/mgg/image/globe_tiles.gif), but at full size. 

I've figured out how to convert a tile into a PNG in what seems to be an accurate manner:

    magick -size 10800x4800 -depth 16 gray:a10g.raw -evaluate AddModulus 50% -auto-level a10g.png

But this produces grayscale output, which is not useful to me. I have tried searching for a way to incorporate the included palette file so that the image is colored properly, but I haven't found a solid answer. Is there a way to do this in ImageMagick? If not, does anyone know of another program compatible with Unix-likes that could do this?

**Comment by fmw42** [comment]
Provide a small example image and the palette file or data.  What is your Imagemagick version and platform/OS?

**Comment by snibgo** [comment, marked as answer]
In your command ...
> magick -size 10800x4800 -depth 16 gray:a10g.raw -evaluate AddModulus 50% -auto-level a10g.png
... Why do you "`-auto-level`"? This loses the fact that "zero" elevation sea-level (or geoid or whatever) is at 50%, and also wrecks the colour-mapping that you want. So remove "`-auto-level`".

The ACT files seem to be headerless RGB 8-bit/channel files, with "zero" elevation in pixel 0, with negative values at the end of the file. So we need to "-roll" it.
```
magick -size 256x1 -depth 8 rgb:glelevpc.act -roll +128+0 map.png
```
Perhaps the "128" should be "127" or "129".

Now we can use map.png as a clut:
```
magick b10g.png map.png -clut -resize 600 x.jpg
```
![x](https://user-images.githubusercontent.com/33812368/200648455-c96a4ef9-349d-4a35-9269-fc0e26fa2714.jpg)

I resize and write to JPG to save bandwidth.

**Reply by chickazee4** [reply]
This works perfectly, thank you!
